[LI-HOTFIX] Skip shutdown safety check at epoch already checked

ambroff opened the pull request:
TICKET = N/A
LI_DESCRIPTION =
In c337e4752 (PR #103) a shutdown safety check was introduced. In
testing this in a large cluster, we found a problem. If the controller
sees that brokerId 1 is safe to shut down at brokerEpoch 99, it will
add that broker to `controllerContext.shuttingDownBrokerIds`. But it
might still not allow shutdown because some leadership transitions
must happen first.

In the process of performing those leadership transitions, brokerId 1
may receive StopReplica, which would cause it to leave the
ISR. Subsequent ControlledShutdownRequest attempts would then fail the
`safeToShutdown` safety check, so the broker would never be allowed to
shut down.

This patch just adjust the implementation of `safeToShutdown`
slightly. Now, if the controller has said that a broker is safe to
shut down at a given epoch, it will always allow that shutdown for
that epoch without performing any additional checks. This allows the
full controlled shutdown process to complete which in a large, busy
cluster may take several rounds of `ControlledShutdownRequest`.

EXIT_CRITERIA = N/A

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ambroff committed:
[LI-HOTFIX] Skip shutdown safety check at epoch already checked

TICKET = N/A
LI_DESCRIPTION =
In c337e4752 (PR #103) a shutdown safety check was introduced. In
testing this in a large cluster, we found a problem. If the controller
sees that brokerId 1 is safe to shut down at brokerEpoch 99, it will
add that broker to `controllerContext.shuttingDownBrokerIds`. But it
might still not allow shutdown because some leadership transitions
must happen first.

In the process of performing those leadership transitions, brokerId 1
may receive StopReplica, which would cause it to leave the
ISR. Subsequent ControlledShutdownRequest attempts would then fail the
`safeToShutdown` safety check, so the broker would never be allowed to
shut down.

This patch just adjust the implementation of `safeToShutdown`
slightly. Now, if the controller has said that a broker is safe to
shut down at a given epoch, it will always allow that shutdown for
that epoch without performing any additional checks. This allows the
full controlled shutdown process to complete which in a large, busy
cluster may take several rounds of `ControlledShutdownRequest`.

EXIT_CRITERIA = N/A

gitlw commented:
Thanks for the patch. The patch looks good to me for addressing the described scenario.

There is another related scenario:
Say a topic has RF=4, min ISR=2, and redundancy factor of 2, if the requesting broker is a lagging follower and all the other 3 replicas are in the ISR, the current implementation will block the request since it requires all 4 replicas to be in the ISR.
I feel it's better to exclude the requesting broker and ensure there are no partitions of "a remaining set size < min ISR + redundancy - 1". Thoughts?

ambroff commented:
Thanks Lucas!
 
> Say a topic has RF=4, min ISR=2, and redundancy factor of 2, if the requesting broker is a lagging follower and all the other 3 replicas are in the ISR, the current implementation will block the request since it requires all 4 replicas to be in the ISR.
> I feel it's better to exclude the requesting broker and ensure there are no violations of "a remaining set size < min ISR + redundancy - 1". Thoughts?

Yeah I thought about that a little bit. The original intent was to block allowing shutdown until the cluster is in a healthy state for any partitions owned by the requesting broker. So before allowing shutdown, there must be min.insync.replicas + redundancy factor replicas in the ISR.

So like you said for your that would be 4 brokers.

At first I didn't like the idea of allowing shutdown when the cluster hadn't fully recovered. But it's true that if a replica is out of sync then shutting it down should have no impact.

I think I like your suggestion. Maybe instead of making the check

```
liveReplicasInIsr < (minISR + config.controlledShutdownSafetyCheckRedundancyFactor - 1)
```

We should just change the definition of the redundancy factor property. Instead of saying that `|ISR| must be >= min.insync.replicas + redundancyFactor`, it should be, that removing teh broker should leave `min.insync.replicas + redundancyFactor` replicas in the ISR.

```
debug(s"$partition has min.insync.replicas=$minISR and a redundancy factor of ${config.controlledShutdownSafetyCheckRedundancyFactor}. Removing broker $id will leave $remainingLiveReplicasInIsr live replicas in the ISR.")
remainingLiveReplicasInIsr < (minISR + config.controlledShutdownSafetyCheckRedundancyFactor)
```

ambroff committed:
Feedback from Lucas

He pointed out that if a broker which is attempting to shutdown is
already out of ISR, then it is probably pretty harmless to allow it to
shut down if the safety check would have allowed it to when there are
no under-replicated partitions.

So this changes the logic slightly.

ControlledShutdownSafetyCheckEnableProp is now defined as, "Shutdown
will be allowed if removing the candidate broker will leave
min.insync.replicas + ControlledShutdownSafetyCheckEnableProp replicas
in the ISR.

The default redundancy factor is changed to 1.

gitlw reviewed (approved):
LGTM, thanks for the fix!
